### PR TITLE
Implement Series.combine

### PR DESCRIPTION
**Pull request by itholic**
Implement [Series.combine](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.combine.html#pandas.Series.combine)

- Basic example
```python
>>> s1 = ks.Series({'falcon': 330.0, 'eagle': 160.0})
>>> s1
falcon    330.0
eagle     160.0
Name: 0, dtype: float64
>>> s2 = ks.Series({'falcon': 345.0, 'eagle': 200.0, 'duck': 30.0})
>>> s2
falcon    345.0
eagle     200.0
duck       30.0
Name: 0, dtype: float64

>>> s1.combine(s2, max).sort_index()
duck        NaN
eagle     200.0
falcon    345.0
Name: 0, dtype: float64

>>> s1.combine(s2, max, fill_value=0).sort_index()
duck       30.0
eagle     200.0
falcon    345.0
Name: 0, dtype: float64

>>> s1.combine(s2, max, fill_value=50).sort_index()
duck       50.0
eagle     200.0
falcon    345.0
Name: 0, dtype: float64
```

- MultiIndex
```python
>>> midx1 = pd.MultiIndex([['lama', 'cow', 'falcon', 'koala'],
...                        ['speed', 'weight', 'length', 'power']],
...                       [[0, 3, 1, 1, 1, 2, 2, 2],
...                        [0, 2, 0, 3, 2, 0, 1, 3]])
>>> midx2 = pd.MultiIndex([['lama', 'cow', 'falcon'],
...                        ['speed', 'weight', 'length']],
...                       [[0, 0, 0, 1, 1, 1, 2, 2, 2],
...                        [0, 1, 2, 0, 1, 2, 0, 1, 2]])
>>> kser1 = ks.Series([45, 200, 1.2, 30, 250, 1.5, 320, 1], index=midx1)
>>> kser2 = ks.Series([-45, 200, -1.2, 30, -250, 1.5, 320, 1, -0.3], index=midx2)

>>> kser1
lama    speed      45.0
koala   length    200.0
cow     speed       1.2
        power      30.0
        length    250.0
falcon  speed       1.5
        weight    320.0
        power       1.0
Name: 0, dtype: float64
>>> kser2
lama    speed     -45.0
        weight    200.0
        length     -1.2
cow     speed      30.0
        weight   -250.0
        length      1.5
falcon  speed     320.0
        weight      1.0
        length     -0.3
Name: 0, dtype: float64

>>> kser1.combine(kser2, max).sort_index()
cow     length    250.0
        power      30.0
        speed      30.0
        weight      NaN
falcon  length      NaN
        power       1.0
        speed     320.0
        weight    320.0
koala   length    200.0
lama    length      NaN
        speed      45.0
        weight      NaN
Name: 0, dtype: float64
```

- with scala value
```python
>>> kser1.combine(100, max)
cow     length    250.0
falcon  weight    320.0
cow     speed     100.0
falcon  speed     100.0
lama    speed     100.0
koala   length    200.0
cow     power     100.0
falcon  power     100.0
Name: 0, dtype: float64
```

### CODE REVIEW

**Comment by itholic** [comment]
Performing with scala values is failing at pandas<1.0.0 (it seems bug of pandas)

- pandas<1.0.0
```python
>>> pser1
eagle     160.0
falcon    330.0
Name: 0, dtype: float64

>>> pser1.combine(200, max)
Traceback (most recent call last):
...
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

- pandas >= 1.0.0
```python
>>> pser1
falcon    330.0
eagle     160.0
Name: 0, dtype: float64

>>> pser1.combine(200, max)
falcon    330.0
eagle     200.0
Name: 0, dtype: float64
```

File: databricks/koalas/series.py [hunk 1]
```diff
@@ -4324,6 +4325,149 @@ def pct_change(self, periods=1):
...
+            result = func(x, y)
+            if not isinstance(result, (int, float)):
+                raise NotImplementedError(
+                    "Only `func` with returning numeric-type is supported for now."
+                )
+            return float(result)
+
+        name = self.name
+        this = "__this_0"
+        that = "__that_0"
```
Comment (HyukjinKwon) [inline databricks/koalas/series.py]: Here too, seems it assumes the column name is always `0`.

File: databricks/koalas/series.py [hunk 2]
```diff
@@ -4324,6 +4325,149 @@ def pct_change(self, periods=1):
...
+
+        if isinstance(other, ks.Series):
+            combined = combine_frames(self.to_frame(), other)
+        else:
+            # Managing scala value
```
Comment (HyukjinKwon) [inline databricks/koalas/series.py]: `scala` -> `scalar`

File: databricks/koalas/series.py [hunk 3]
```diff
@@ -4324,6 +4325,149 @@ def pct_change(self, periods=1):
...
+        eagle     200.0
+        falcon    345.0
+        Name: 0, dtype: float64
+
+        >>> reset_option("compute.ops_on_diff_frames")
+        """
+        # TODO: only `func` with return numeric-type is supported for now,
+        # assuming this case is the most frequent.
+        # because we can't infer the return type of `func` before it actually return.
+        @udf("double")
```
Comment (HyukjinKwon) [inline databricks/koalas/series.py]: We should not use the regular `udf`. This is super slow (can be 100+ times slower then other expressions).

File: databricks/koalas/series.py [hunk 4]
```diff
@@ -4324,6 +4325,149 @@ def pct_change(self, periods=1):
...
+        def spark_udf(x, y):
```
Comment (HyukjinKwon) [inline databricks/koalas/series.py]: I think you can use `isnan` with `when` and `otherwise` expressions.

File: databricks/koalas/series.py [hunk 5]
```diff
@@ -4324,6 +4325,149 @@ def pct_change(self, periods=1):
...
+            x = float("NaN") if x is None else x
+            y = float("NaN") if y is None else y
+            result = func(x, y)
```
Comment (HyukjinKwon) [inline databricks/koalas/series.py]: We should probably is pandas UDF

**Comment by xinrong-databricks** [comment]
Created https://issues.apache.org/jira/browse/SPARK-36402 for porting this :)

**Pull request closed without merging**